resolver/resolver.go

```diff
@@ -457,7 +457,12 @@ func isUDP(w dns.ResponseWriter) bool {
 // truncate removes answers until the given dns.Msg fits the permitted
 // length of the given transmission channel and sets the TC bit.
 // See https://tools.ietf.org/html/rfc1035#section-4.2.1
-func truncate(m *dns.Msg, udp bool) *dns.Msg {
+// If `setTruncateBit` is true, the Truncate bit will be set if the message
+// has been truncated already or gets truncated here.
+// If `setTruncateBit` is false, the message will have its Truncate bit
+// cleared if it is already set, and won't set it even if it does get
+// truncated.
+func truncate(m *dns.Msg, udp bool, setTruncateBit bool) *dns.Msg {
```

Comment:
I know the original function had the same signature, but I feel like we shouldn't return the `*dns.Msg`. It's kinda misleading, as we mutate the passed in `m *dns.Msg` and then return the exact same pointer to it. Passing something in and getting returned another implies that the original is left untouched and a completely new thing is returned, which is not the case here.

This is also causing what I believe to be a bug in the unit tests, I have a suspicion that all the times when the input and output of `truncate` is compared, it will always trivially be true because they are the same. Instead, what should be done is to _start_ with 2 separate messages, created the exact same way, mutating one, and _then_ comparing them.